File: src/main/kotlin/com/draw/drawmusic/track/TrackBar.java
package com.draw.drawmusic.track;

import com.draw.drawmusic.controllers.FXMLController;
import com.draw.drawmusic.history.TrackHistory;
import com.draw.drawmusic.tools.ArrayTool;
import com.draw.drawmusic.tools.KeyBoardTool;
import com.draw.drawmusic.tools.Order;
import javafx.animation.FadeTransition;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.ScrollPane;
import javafx.scene.control.ToolBar;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyCombination;
import javafx.scene.layout.VBox;
import javafx.util.Duration;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;

public class TrackBar {
    public static final int TRACK_MAX_NUMBER = 50;
    private static ToolBar toolBar;
    private static VBox contentBar;
    private static ScrollPane scrollPane;
    private static Label displayTrackNumbers;
    private static Button addButton;
    private static final ArrayList<Track> trackElements = new ArrayList<>();

    public static void init(VBox _trackBarContent, ScrollPane _trackBarScrollPane, ToolBar _toolBar) {
        contentBar = _trackBarContent;
        scrollPane = _trackBarScrollPane;
        toolBar = _toolBar;

        fitScrollPaneSize();
        setToolBar();
    }

    private static void fitScrollPaneSize() {
        scrollPane.setFitToWidth(true);
        scrollPane.setFitToHeight(true);
    }

    private static void setToolBar() {
        matchToolBarItems();
        setToolBarEvent();
    }

    public static void matchKeyboardShortcuts() {
        KeyBoardTool.setKeyEventHandler(TrackBar::deleteSelectedElements, KeyCode.DELETE, KeyCombination.SHIFT_DOWN);
        KeyBoardTool.setKeyEventHandler(TrackBar::addButtonClicked, KeyCode.N, KeyCombination.CONTROL_DOWN);
    }

    public static void addButtonClicked() {
        loadTrackElementFXML();
        TrackHistory.saveAddHistory(ArrayTool.last(trackElements));
    }

    public static void addElement(Track newElement) {
        trackElements.add(newElement);
        syncOrderInDisplay();
        selectElements(newElement, false, false);
        displayAndUpdateElements();
        fadeInElement(newElement);
        scrollDownPossible();
        setDisplayTrackNumbers();
    }

    private static void displayAndUpdateElements() {
        displayElements();
        updateElements();
    }

    public static void selectAndUpdateElements(Track element) {
        selectElements(element, false, false);
        updateElements();
    }

    public static void selectAndUpdateElements(Track element, boolean ctrlClicked, boolean shiftClicked) {
        selectElements(element, ctrlClicked, shiftClicked);
        updateElements();
    }

    private static void deleteSelectedElements() {
        for(Track element : getSelectedList()) {
            deleteElement(element);
        }
    }

    public static void deleteElement(Track track) {
        selectAndUpdateElements(ArrayTool.closeElement(trackElements, track));
        trackElements.remove(track);
        setDisplayTrackNumbers();
        displayElements();
    }

    public static void moveUp(Track a) {
        switchTwoTracks(a, ArrayTool.previous(trackElements, a));
    }

    public static void moveDown(Track a) {
        switchTwoTracks(a, ArrayTool.next(trackElements, a));
    }

    public static Order makeLastOrder() {
        if(trackElements.isEmpty()) return Order.DEFAULT;
        return Order.nextOrderThan(ArrayTool.last(trackElements).orderInTrackBar);
    }

    private static void switchTwoTracks(Track a, Track b) {
        Order.swap(a.orderInTrackBar, b.orderInTrackBar);
        syncOrderInDisplay();
        displayElements();
    }

    private static void syncOrderInDisplay() {
        Collections.sort(trackElements);
    }

    private static void matchToolBarItems() {
        addButton = (Button) ArrayTool.first(toolBar.getItems());
        displayTrackNumbers = (Label) ArrayTool.last(toolBar.getItems());
        displayTrackNumbers.setText("0/" + TRACK_MAX_NUMBER);
    }

    private static void setToolBarEvent() {
        addButton.setOnAction(actionEvent -> addButtonClicked());
    }

    private static void loadTrackElementFXML() {
        FXMLController.Companion.fxmlLoad("midi-track.fxml");
    }

    @NotNull
    private static ArrayList<Track> getSelectedList() {
        final ArrayList<Track> selected = new ArrayList<>();
        for(Track element : trackElements) {
            if(element.trackSelect != TrackSelect.unSelected) {
                selected.add(element);
            }
        }
        return selected;
    }

    private static void setDisplayTrackNumbers() {
        displayTrackNumbers.setText(trackElements.size() + "/" + TRACK_MAX_NUMBER);
    }

    private static void displayElements() {
        contentBar.getChildren().clear();
        for (Track element: trackElements) {
            contentBar.getChildren().add(element.trackElement.getGridPane());
        }
    }

    private static void updateElements() {
        for (Track element: trackElements) {
            element.trackElement.updateShape();
        }
    }

    private static void fadeInElement(@NotNull Track node) {
        FadeTransition ft = new FadeTransition(Duration.millis(500), node.trackElement.getGridPane());
        ft.setFromValue(0);
        ft.setToValue(1);
        ft.play();
    }

    private static void scrollDownPossible() {
        scrollPane.setVvalue(scrollPane.getVmax());
    }

    private static void selectElements(Track element, boolean ctrlClicked, boolean shiftClicked) {
        if (shiftClicked) {
            int index = trackElements.indexOf(element), siz = trackElements.size();
            int start = -1;
            for(int i = 0; i < siz; i++) {
                if(trackElements.get(i).trackSelect == TrackSelect.currentSelected) {
                    start = i;
                }
            }
            if(start == -1) {
                element.trackSelect = TrackSelect.currentSelected;
            } else {
                for(Track _element : trackElements)
                    _element.trackSelect = TrackSelect.unSelected;
                for(int i = Integer.min(index, start); i <= Integer.max(index, start); i++) {
                    if(i == start) trackElements.get(i).trackSelect = TrackSelect.currentSelected;
                    else trackElements.get(i).trackSelect = TrackSelect.multiSelected;
                }
            }
        } else if (ctrlClicked) {
            if(element.trackSelect != TrackSelect.unSelected) {
                element.trackSelect = TrackSelect.unSelected;
            } else {
                for (Track _element : trackElements) {
                    if (_element.trackSelect == TrackSelect.currentSelected)
                        _element.trackSelect = TrackSelect.multiSelected;
                }
                element.trackSelect = TrackSelect.currentSelected;
            }
        } else {
            for (Track _element : trackElements) {
                _element.trackSelect = TrackSelect.unSelected;
            }
            element.trackSelect = TrackSelect.currentSelected;
        }
    }
}

File: src/main/kotlin/com/draw/drawmusic/track/TrackBar.kt
package com.draw.drawmusic.track

import com.draw.drawmusic.controllers.FXMLController.Companion.fxmlLoad
import com.draw.drawmusic.history.TrackHistory
import com.draw.drawmusic.tools.ArrayTool
import com.draw.drawmusic.tools.KeyBoardTool.setKeyEventHandler
import com.draw.drawmusic.tools.Order
import javafx.animation.FadeTransition
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.scene.control.Button
import javafx.scene.control.Label
import javafx.scene.control.ScrollPane
import javafx.scene.control.ToolBar
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyCombination
import javafx.scene.layout.VBox
import javafx.util.Duration

// TODO: 트랙 삭제 후 ctrl-z 하면 에러

// TODO: 트랙 삭제 후 ctrl-z 하면 에러
object TrackBar {
    private const val TRACK_MAX_NUMBER = 50
    private var toolBar: ToolBar? = null
    private var contentBar: VBox? = null
    private var scrollPane: ScrollPane? = null
    private var displayTrackNumbers: Label? = null
    private var addButton: Button? = null
    private val trackElements = ArrayList<Track>()
    private val selectedList: ArrayList<Track>
        get() {
            val selected = ArrayList<Track>()
            for (element in trackElements) {
                if (element.trackSelect !== TrackSelect.UnSelected) {
                    selected.add(element)
                }
            }
            return selected
        }

    fun init(_trackBarContent: VBox?, _trackBarScrollPane: ScrollPane?, _toolBar: ToolBar?) {
        contentBar = _trackBarContent
        scrollPane = _trackBarScrollPane
        toolBar = _toolBar
        fitScrollPaneSize()
        setToolBar()
    }

    private fun fitScrollPaneSize() {
        scrollPane!!.isFitToWidth = true
        scrollPane!!.isFitToHeight = true
    }

    private fun setToolBar() {
        matchToolBarItems()
        setToolBarEvent()
    }

    fun matchKeyboardShortcuts() {
        setKeyEventHandler(
            TrackBar::deleteSelectedElements,
            KeyCode.DELETE,
            KeyCombination.SHIFT_DOWN
        )

        setKeyEventHandler(
            TrackBar::addButtonClicked,
            KeyCode.N,
            KeyCombination.CONTROL_DOWN
        )
    }

    private fun addButtonClicked() {
        loadTrackElementFXML()
        TrackHistory.saveAddHistory(ArrayTool.last(trackElements))
    }

    fun addElement(newElement: Track) {
        trackElements.add(newElement)
        selectElements(newElement, ctrlClicked = false, shiftClicked = false)
        displayAndUpdateElements()
        fadeInElement(newElement)
        scrollDown()
        setDisplayTrackNumbers()
        for (i in trackElements) {
            println(i.toString() + " : " + i.orderInTrackBar)
        }
        println("====================")
    }

    private fun displayAndUpdateElements() {
        displayElements()
        updateElements()
    }

    private fun selectAndUpdateElements(element: Track) {
        selectElements(element, ctrlClicked = false, shiftClicked = false)
        updateElements()
    }

    fun selectAndUpdateElements(element: Track, ctrlClicked: Boolean, shiftClicked: Boolean) {
        selectElements(element, ctrlClicked, shiftClicked)
        updateElements()
    }

    private fun deleteSelectedElements() {
        for (element in selectedList) {
            deleteElement(element)
        }
    }

    fun deleteElement(track: Track) {
        selectAndUpdateElements(ArrayTool.closeElement(trackElements, track))
        trackElements.remove(track)
        setDisplayTrackNumbers()
        displayElements()
    }

    fun moveUp(a: Track) {
        switchTwoTracks(a, ArrayTool.previous(trackElements, a))
    }

    fun moveDown(a: Track) {
        switchTwoTracks(a, ArrayTool.next(trackElements, a))
    }

    fun makeLastOrder(): Order {
        return if (trackElements.isEmpty()) Order.DEFAULT else Order.nextOrderThan(ArrayTool.last(trackElements).orderInTrackBar)
    }

    private fun switchTwoTracks(a: Track, b: Track) {
        Order.swap(a.orderInTrackBar, b.orderInTrackBar)
        syncOrderInDisplay()
        displayElements()
    }

    private fun syncOrderInDisplay() {
        trackElements.sort()
    }

    private fun matchToolBarItems() {
        addButton = ArrayTool.first(toolBar!!.items) as Button
        displayTrackNumbers = ArrayTool.last(toolBar!!.items) as Label
        displayTrackNumbers!!.text = "0/$TRACK_MAX_NUMBER"
    }

    private fun setToolBarEvent() {
        addButton!!.onAction = EventHandler { _: ActionEvent? -> addButtonClicked() }
    }

    private fun loadTrackElementFXML() {
        fxmlLoad("midi-track.fxml")
    }

    private fun setDisplayTrackNumbers() {
        displayTrackNumbers!!.text = trackElements.size.toString() + "/" + TRACK_MAX_NUMBER
    }

    private fun displayElements() {
        contentBar!!.children.clear()
        for (element in trackElements) {
            contentBar!!.children.add(element.trackElement.gridPane)
        }
    }

    private fun updateElements() {
        for (element in trackElements) {
            element.trackElement.updateShape()
        }
    }

    private fun fadeInElement(node: Track) {
        val ft = FadeTransition(Duration.millis(500.0), node.trackElement.gridPane)
        ft.fromValue = 0.0
        ft.toValue = 1.0
        ft.play()
    }

    private fun scrollDown() {
        scrollPane!!.vvalue = scrollPane!!.vmax
    }

    private fun selectElements(element: Track, ctrlClicked: Boolean, shiftClicked: Boolean) {
        if (shiftClicked) {
            val index = trackElements.indexOf(element)
            val siz = trackElements.size
            var start = -1
            for (i in 0 until siz) {
                if (trackElements[i].trackSelect === TrackSelect.CurrentSelected) {
                    start = i
                }
            }
            if (start == -1) {
                element.trackSelect = TrackSelect.CurrentSelected
            } else {
                for (_element in trackElements) _element.trackSelect = TrackSelect.UnSelected
                for (i in Integer.min(index, start)..Integer.max(index, start)) {
                    if (i == start) trackElements[i].trackSelect =
                        TrackSelect.CurrentSelected else trackElements[i].trackSelect = TrackSelect.MultiSelected
                }
            }
        } else if (ctrlClicked) {
            if (element.trackSelect !== TrackSelect.UnSelected) {
                element.trackSelect = TrackSelect.UnSelected
            } else {
                for (_element in trackElements) {
                    if (_element.trackSelect === TrackSelect.CurrentSelected) _element.trackSelect =
                        TrackSelect.MultiSelected
                }
                element.trackSelect = TrackSelect.CurrentSelected
            }
        } else {
            for (_element in trackElements) {
                _element.trackSelect = TrackSelect.UnSelected
            }
            element.trackSelect = TrackSelect.CurrentSelected
        }
    }
}
